### DISCUSSION TITLE
How can I use Phantom Camera in a first-person game with cutscenes?

### DISCUSSION THREAD

**Discussion by stevenmraines** [ramokz/phantom-camera#361] [Question / Help Wanted] [answered]
I'm working on a first-person perspective game. My player scene's root is a CharacterBody3D. It has a child Camera3D which is controlled by the player script to look around with the mouse. My first "level" of the game has a few cutscenes in which I've animated several other Camera3Ds.

I want to use Phantom Camera to make life a little easier, and I've been playing around with it and learning how to manage multiple PCams. The problem is that, as far as I can tell, Phantom Camera needs there to be exactly one Camera3D with a child of type PhantomCameraHost at the first level of the scene tree. So how can I handle the player scene having it's own Phantom Camera and switch between it and the cutscene cameras?

I've tried giving the player an exported var called p_cam, then adding a PCam to my level and assigning it in the p_cam slot in the editor. And I have an animation that plays when the level is started to control the priorities of the various cameras. But when I start my level I just get a grey screen.

It feels like I'm using Phantom Camera incorrectly, like it's only meant for controlling a 3rd person camera and not like what I'm trying to do. Is there a way to make it work in a first-person game with cutscenes?

**Comment by stevenmraines** [comment, marked as answer]
I think I may have found a solution. I removed any Camera3Ds and PCams from my player scene and added a Node3D called PCamPlaceholder where the Camera3D was originally positioned, with the same orientation. I added this line to the player script to get a reference to it:

```
@onready var p_cam_placeholder := $PCamPlaceholder
```

In my level scene I added a PhantomCamera3D called PlayerPhantomCamera3D. For this player PCam I set follow mode to glued and look at mode to mimic, and I set the tween duration to 0. Now in my level script I set everything up in the _ready function:

```
@onready var player := $Player

var player_p_cam : PhantomCamera3D

func _ready():
	for cam in PhantomCameraManager.phantom_camera_3ds:
		if cam.name == "PlayerPhantomCamera3D":
			player_p_cam = cam
			player_p_cam.look_at_target = player.p_cam_placeholder
			player_p_cam.follow_target = player.p_cam_placeholder
```

And with that, everything seems to be working. When I want to animate cameras in some cutscene I use this method:

```
func switch_cameras(cam_path : NodePath):
	for cam in PhantomCameraManager.phantom_camera_3ds:
		var node_name = cam_path.get_name(cam_path.get_name_count() - 1)
		var priority := 1 if cam.name == node_name else 0
		cam.set_priority(priority)
```

And when I want to end the cutscene I use this method:

```
func end_cutscene():
	for cam in PhantomCameraManager.phantom_camera_3ds:
		cam.set_priority(0)
	
	player_p_cam.set_priority(1)
```

**Reply by ramokz** [reply]
Glad you figured it out :slightly_smiling_face: 

Yes, having the `Camera3D` nested inside the player scene adds a lot of complication to the setup, so what you changed is what I would have suggested as well.
The only other thing worth mentioning is that you could retain the `PCam` inside the player scene still if you would prefer keeping it in one place. Obviously, you can't preview it from the player scene itself and would need to do so from the main scene where the `Camera3D` is. So this is only really for organisational and player scene reusability purposes.

**Reply by stevenmraines** [reply]
I had tried keeping the camera in the player scene and I couldn't get it to work for some reason. IIRC I was just getting a grey screen when I loaded my scene. But I hadn't thought to use the glued/mimic follow and look at modes then, so maybe I'll give it another go, thanks.

**Reply by ramokz** [reply]
That's strange.

You obviously shouldn't be able to see anything if you just run the player scene, as it's missing the `Camera` node. But the main scene should work as is if it contains both the `Camera` and player scene, with the `PCam` inside of that.

Tested it just now myself without setting any properties to the `PCam` node.

**Reply by stevenmraines** [reply]
Yeah, not sure what the problem was before, but it seems to be working now. Thanks again!
